### PR TITLE
Introduced MessageBuilder to simplify the creation of MQTT messages

### PR DESCRIPTION
**Pull request by andsel**
Motivation:
This PR would simplify the creation of MqttMessages in tests, to avoid long parameters list in constructors of variable, fixed and payload parts.

Modification:

Just introduced a builder class that permit to create messaged with chined notation.

Result:

Created MessageBuilder class with some factory methods (for connect, publish etc) and some inner classes used to carry temporary variables. At the end there  is a build method that contains the logic to wire up the parts of the message.

### CODE REVIEW

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MessageBuilder.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    public static class PublishBuilder {
+        private String topic;
+        private boolean retained;
+        private MqttQoS qos;
+        private byte[] payload;
+        private int messageId;
+
+        public PublishBuilder topicName(String topic) {
+            this.topic = topic;
+            return PublishBuilder.this;
```
Comment (netkins) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MessageBuilder.java]: MAJOR Useless qualified this usage in the same class. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUselessQualifiedThis)

**Comment by andsel** [comment]
Hi @normanmaurer let me know if there is something bad or the PR is not good.
Thanks 
 Andrea

**Comment by normanmaurer** [comment]
@andsel sorry a bit busy... will review soon in detail

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MessageBuilder.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+ * under the License.
+ */
+package io.netty.handler.codec.mqtt;
+
+import io.netty.buffer.Unpooled;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class MessageBuilder {
```
Comment (louxiu) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MessageBuilder.java]: `MessageBuilder` is a bit general. Would `MqttMessageBuilder` is better than `MessageBuilder`?

**Comment by andsel** [comment]
Thanks @louxiu for the hint, done it

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2014 The Netty Project
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: 2017

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+ * under the License.
+ */
+package io.netty.handler.codec.mqtt;
+
+import io.netty.buffer.Unpooled;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class MqttMessageBuilder {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: As this class contains multiple builders consider naming it `MqttMessageBuilders`.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+    public static class PublishBuilder {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: final

Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: also package-private constructor

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        public MqttUnsubscribeMessage build() {
+            MqttFixedHeader mqttFixedHeader =
+                    new MqttFixedHeader(MqttMessageType.UNSUBSCRIBE, false, MqttQoS.AT_LEAST_ONCE, false, 0);
+            MqttMessageIdVariableHeader mqttVariableHeader = MqttMessageIdVariableHeader.from(messageId);
+            MqttUnsubscribePayload mqttSubscribePayload = new MqttUnsubscribePayload(topicFilters);
+            return new MqttUnsubscribeMessage(mqttFixedHeader, mqttVariableHeader, mqttSubscribePayload);
+        }
+    }
+
+    public static class ConnAckBuilder {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: package-private constructor and final

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        public MqttSubscribeMessage build() {
+            MqttFixedHeader mqttFixedHeader =
+                    new MqttFixedHeader(MqttMessageType.SUBSCRIBE, false, MqttQoS.AT_LEAST_ONCE, false, 0);
+            MqttMessageIdVariableHeader mqttVariableHeader = MqttMessageIdVariableHeader.from(messageId);
+            MqttSubscribePayload mqttSubscribePayload = new MqttSubscribePayload(subscriptions);
+            return new MqttSubscribeMessage(mqttFixedHeader, mqttVariableHeader, mqttSubscribePayload);
+        }
+    }
+
+    public static class UnsubscribeBuilder {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: package-private constructor and final

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+                            willFlag,
+                            cleanSession,
+                            keepAliveSecs);
+            MqttConnectPayload mqttConnectPayload =
+                    new MqttConnectPayload(clientId, willTopic, willMessage, username, password);
+            return new MqttConnectMessage(mqttFixedHeader, mqttConnectVariableHeader, mqttConnectPayload);
+        }
+    }
+
+    public static class SubscribeBuilder {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: package-private constructor and final

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        }
+
+        public MqttPublishMessage build() {
+            MqttFixedHeader mqttFixedHeader = new MqttFixedHeader(MqttMessageType.PUBLISH, false, qos, retained, 0);
+            MqttPublishVariableHeader mqttVariableHeader = new MqttPublishVariableHeader(topic, messageId);
+            return new MqttPublishMessage(mqttFixedHeader, mqttVariableHeader, Unpooled.buffer().writeBytes(payload));
+        }
+    }
+
+    public static class ConnectBuilder {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: package-private constructor and final

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+                            keepAliveSecs);
+            MqttConnectPayload mqttConnectPayload =
+                    new MqttConnectPayload(clientId, willTopic, willMessage, username, password);
+            return new MqttConnectMessage(mqttFixedHeader, mqttConnectVariableHeader, mqttConnectPayload);
+        }
+    }
+
+    public static class SubscribeBuilder {
+
+        private List<MqttTopicSubscription> subscriptions = new ArrayList<MqttTopicSubscription>();
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: consider lazy instance it to eliminate unessary allocations when no subscriptions are added.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java (hunk 5 above)
Comment (andsel) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilder.java]: Hi @normanmaurer may I ask you why make the package-private the MqttMessageBuilder constructor and not private? Beeing private no other classes can instantiate it

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+                    new MqttFixedHeader(MqttMessageType.SUBSCRIBE, false, MqttQoS.AT_LEAST_ONCE, false, 0);
+            MqttMessageIdVariableHeader mqttVariableHeader = MqttMessageIdVariableHeader.from(messageId);
+            MqttSubscribePayload mqttSubscribePayload = new MqttSubscribePayload(subscriptions);
+            return new MqttSubscribeMessage(mqttFixedHeader, mqttVariableHeader, mqttSubscribePayload);
+        }
+    }
+
+    public static final class UnsubscribeBuilder {
+
+        private List<String> topicFilters = new ArrayList<String>();
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java]: @andsel should be final or created lazy when a `topicFilter` is added. Also consider if the default initialCapacity makes sense.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+    public static final class SubscribeBuilder {
+
+        private List<MqttTopicSubscription> subscriptions;
+        private int messageId;
+
+        SubscribeBuilder() {}
+
+        public SubscribeBuilder addSubscription(MqttQoS qos, String topic) {
+            if (subscriptions == null) {
+                subscriptions = new ArrayList<MqttTopicSubscription>();
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java]: does it make sense to use a different initialCapacity here ?

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+            this.willRetain = willRetain;
+            return this;
+        }
+
+        public ConnectBuilder hasUser() {
+            this.hasUser = true;
+            return this;
+        }
+
+        public ConnectBuilder hasPassword() {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java]: Does it make sense to pass boolean here to be consistent with other methods ?

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+            this.willMessage = willMessage;
+            return this;
+        }
+
+        public ConnectBuilder willRetain(boolean willRetain) {
+            this.willRetain = willRetain;
+            return this;
+        }
+
+        public ConnectBuilder hasUser() {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java]: Does it make sense to pass boolean here to be consistent with other methods ?

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+            this.retained = retained;
+            return this;
+        }
+
+        public PublishBuilder qos(MqttQoS qos) {
+            this.qos = qos;
+            return this;
+        }
+
+        public PublishBuilder payload(byte[] payload) {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java]: I think this should take a `ByteBuf`.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+        private boolean sessionPresent;
+
+        ConnAckBuilder() {}
+
+        public ConnAckBuilder returnCode(MqttConnectReturnCode returnCode) {
+            this.returnCode = returnCode;
+            return this;
+        }
+
+        public ConnAckBuilder sessionPresent() {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageBuilders.java]: Does it make sense to pass boolean here to be consistent with other methods ?

**Comment by andsel** [comment]
Hi @normanmaurer fixed with your suggestions plus changed password() and username() methods to also set the corresponding flag.

**Comment by normanmaurer** [comment]
@Scottmitch PTAL as well.

**Comment by normanmaurer** [comment]
@andsel please squash and sign our ICLA (http://netty.io/s/icla) . After that I will pull in.

**Comment by andsel** [comment]
Hi @normanmaurer  I've aligned my local copy to do a "rebase -i " to squash my commits in just one, but probably I missed something, because in this PR now are present a lot of commits that aren't mine. How can we recover?

**Comment by normanmaurer** [comment]
@andsel check "git reflog" and git reset to the correct sha.

**Comment by andsel** [comment]
@normanmaurer I transformed this PR in a mess, this evening I'll close it and create a new one with just a single commit

**Comment by normanmaurer** [comment]
@andsel ok .. Just ping me once done

**Comment by andsel** [comment]
@normanmaurer I close this because I've reated a new clean PR (#6407)

**Pull request closed without merging**